src/BreakpointViewer.cpp

```diff
@@ -0,0 +1,621 @@
+				Breakpoints::Type wtype = static_cast<Breakpoints::Type>(combo->currentIndex() + 1);
+				adrLen = (wtype == Breakpoints::WATCHPOINT_IOREAD || wtype == Breakpoints::WATCHPOINT_IOWRITE)
+					? 2 : 4;
+			} else {
+				adrLen = 4;
+			}
+			int begin, end;
+
+			if (processLocationField(index, type, item->text(), begin, end)) {
+				userMode = false;
+				item->setText(QString("%1%2%3").arg(hexValue(begin, adrLen))
+					.arg(end == begin || end == -1 ? "" : ":")
+					.arg(end == begin || end == -1 ? "" : hexValue(end, adrLen)));
+				userMode = true;
+			} else {
+				userMode = false;
+				enabled = false;
+				item->setText("");
+				enabledItem->setCheckState(Qt::Unchecked);
+				userMode = true;
+			}
+			if (!enabled) return;
+			break;
+		}
+		case SLOT: {
+			qint8 ps, ss;
+			if (processSlotField(index, item->text(), ps, ss)) {
+				userMode = false;
+				item->setText(QString("%1/%2")
+					.arg(ps == -1 ? QChar('X') : QChar('0' + ps))
+					.arg(ss == -1 ? QChar('X') : QChar('0' + ss)));
+				userMode = true;
+			} else {
+				userMode = false;
+				item->setText("X/X");
+				userMode = true;
+			}
+			if (!enabled) return;
+			break;
+		}
+		case SEGMENT: {
+			qint16 segment;
+			if (processSegmentField(index, item->text(), segment)) {
+				userMode = false;
+				item->setText(QString("%1").arg(segment == -1 ? "X" : QString::number(segment)));
+				userMode = true;
+			} else {
+				userMode = false;
+				item->setText("X");
+				userMode = true;
+			}
+			if (!enabled) return;
+			break;
+		}
+		case CONDITION:
+			userMode = false;
+			item->setText(item->text().simplified());
+			userMode = true;
+			if (!enabled) return;
+			break;
+	}
+
+	if (enabled) {
+		if (createBp) {
+			createBreakpoint(type, row);
+		} else {
+			replaceBreakpoint(type, row);
+		}
+	} else if (index != -1) {
+		removeBreakpoint(type, row);
+	} else {
+		assert(index != -1);
+	}
+}
+
+void BreakpointViewer::changeBpTableItem(QTableWidgetItem* item)
+{
+	if (!userMode) return;
+	changeTableItem(BREAKPOINT, item);
+}
+
+void BreakpointViewer::changeWpTableItem(QTableWidgetItem* item)
+{
+	if (!userMode) return;
+	qDebug() << "onWpTableItemChanged() called";
+	changeTableItem(WATCHPOINT, item);
+}
+
+void BreakpointViewer::setBreakpoints(Breakpoints* bps)
+{
+	breakpoints = bps;
+}
+
+void BreakpointViewer::keepUnusedItems()
+{
+	int row = 0;
+
+	while (row < bpTableWidget->rowCount()) {
+		QTableWidgetItem* enabledItem = bpTableWidget->item(row, ENABLED);
+		bool enabled = enabledItem->checkState() == Qt::Checked;
+		if (enabled)
+			bpTableWidget->removeRow(row);
+		else
+			row++;
+	}
+
+	while (row < wpTableWidget->rowCount()) {
+		QTableWidgetItem* enabledItem = wpTableWidget->item(row, ENABLED);
+		bool enabled = enabledItem->checkState() == Qt::Checked;
+		if (enabled)
+			wpTableWidget->removeRow(row);
+		else
+			row++;
+	}
+}
+
+void BreakpointViewer::setRunState() {
+	runState = true;
+}
```

Comment:
(detail)
In most case you follow the style of '{' for a top-level function/method on a new line. But these two break that pattern.